### PR TITLE
[chore][internal/comparetest]Make all exported Compare functions not mutated

### PR DESCRIPTION
**Pull request by jackycsl**
**Description:** <Describe what has changed.>
Make all exported Compare functions not mutated for Logs, Metrics, and Traces.
All the data from the source are copied before any mutation is done on the copied data.

**Link to tracking Issue:** <Issue number if applicable>
#17549

**Testing:** <Describe what testing was performed and which tests were added.>
N/A

**Documentation:** <Describe the documentation added.>
N/A

### CODE REVIEW

File: internal/comparetest/metrics.go [hunk 1]
```diff
@@ -239,14 +251,18 @@ func CompareNumberDataPointSlices(expected, actual pmetric.NumberDataPointSlice)
 // CompareNumberDataPoints compares each part of two given NumberDataPoints and returns
 // an error if they don't match. The error describes what didn't match.
 func CompareNumberDataPoints(expected, actual pmetric.NumberDataPoint) error {
-	if expected.ValueType() != actual.ValueType() {
-		return fmt.Errorf("metric datapoint types don't match: expected type: %s, actual type: %s", numberTypeToString(expected.ValueType()), numberTypeToString(actual.ValueType()))
+	exp, act := pmetric.NewNumberDataPoint(), pmetric.NewNumberDataPoint()
+	expected.CopyTo(exp)
+	actual.CopyTo(act)
```
Comment (dmitryax) [inline internal/comparetest/metrics.go:256]: It's not needed in functions that don't mutate the data like this one. Mutating operation in other functions is sorting

Comment (jackycsl) [reply internal/comparetest/metrics.go:256]: Got it.

File: internal/comparetest/metrics.go [hunk 2]
```diff
@@ -188,21 +196,25 @@ func CompareMetricSlices(expected, actual pmetric.MetricSlice) error {
 // CompareNumberDataPointSlices compares each part of two given NumberDataPointSlices and returns
 // an error if they don't match. The error describes what didn't match.
 func CompareNumberDataPointSlices(expected, actual pmetric.NumberDataPointSlice) error {
-	if expected.Len() != actual.Len() {
-		return fmt.Errorf("number of datapoints does not match expected: %d, actual: %d", expected.Len(), actual.Len())
+	exp, act := pmetric.NewNumberDataPointSlice(), pmetric.NewNumberDataPointSlice()
+	expected.CopyTo(exp)
+	actual.CopyTo(act)
+
```
Comment (jackycsl) [inline internal/comparetest/metrics.go:202]: Should we keep the copied slices here, even though there is no sorting function?

Comment (dmitryax) [reply internal/comparetest/metrics.go:202]: No, it's not needed here

**Review by dmitryax — APPROVED** [review]

**Comment by dmitryax** [comment]
Thanks, @jackycsl

**Review by codeboten — APPROVED** [review]

**Pull request merged**